src/atm.c

```diff
@@ -1,98 +1,205 @@
-			}
-			break;
-
-		case ATM_DEPOSIT:
+	while(tries--) {
 
-			
-			printf("Please enter amount to be deposited. Max of $10,000.00: ");
-			scanf("%d", &amount);
-
-			if (amount > 0 && amount <= MAX_DEPOSIT) {
-				printf("Please Put your cash in an envelope and deposit via slot\n");
-				sleep(5);
-				balance += (double)amount;
-				printf("Your new balance is %lf\n", balance);
-			} else {
-				printf("Invalid amount selected. Please select amount between $%d.00 - $%d.00\n", MIN_DEPOSIT, MAX_DEPOSIT);
-			}
-			
-			break;
+		printf("Please input your PIN: \n");
+		scanf("%s", pin_tries);
+
+		if (!strncmp(pin_tries, pin, sizeof(pin_tries) / sizeof(pin_tries[0])))
+			return ATM_SUCCESS;
+
+		printf("INVALID PIN\n");
+	}
 
-		case ATM_INQUIRY:
+	return ATM_WRONG_PIN;
+}
 
-			printf("Your current balance is %lf\n", balance);
+atm_status_t atm_withdraw(int amount)
+{					
+	if (amount < MIN_WITHDRAW || amount > MAX_WITHDRAW) {
 
-			break;
-		default:
+		printf("Inavlid amount selected.\
+			Please select amount between $%d.00 - $%d.00\n", MIN_WITHDRAW, MAX_WITHDRAW);
+		return ATM_INVALID_AMOUNT;
+	} else if (balance >= (double)amount) {
 
-			printf("Invalid selection\n");
+		balance -= (double)amount;
+		printf("Your new balance is: %lf\n", balance);
+		return ATM_SUCCESS;
+	} else {
 
-			break;
+		printf("You do not have enough funds\n");
+		return ATM_INSUFF_FUNDS;
 	}
 }
 
-/********************************************************************************/
-/*                               PUBLIC FUNCTIONS                               */
-/********************************************************************************/
-void atm_start(void)
+atm_status_t atm_deposit(uint16_t amount)
 {
-	int operation;
+	if (amount > 0 && amount <= MAX_DEPOSIT) {
+
+		printf("Please Put your cash in an envelope and deposit via slot\n");
+		sleep(5);
+		balance += (double)amount;
+		printf("Your new balance is %lf\n", balance);
+		return ATM_SUCCESS; 
+	} else {
+
+		printf("Invalid amount selected.\
+			Please select amount between $%d.00 - $%d.00\n", MIN_DEPOSIT, MAX_DEPOSIT);
+		return ATM_INVALID_AMOUNT;
+	}
+}
 
-	//Get your pin first
+double atm_get_balance(void)
+{
+	return balance;
+}
 
-	printf("%s\nPlease enter your selection: ", welcome_msg);
-	scanf("%d", &operation);
-	operation--;
+bool is_card_inserted(void)
+{
+	//XXX: TODO: How to check if card has been inserted?
+	return card_inserted;
+}
 
-	if (operation < ATM_WITHDRAW || operation > ATM_INQUIRY) {
-		printf("Invalid selection. Goodbye\n");
-		return;
-	}
+static void atm_state_machine(void)
+{
+	//int amount = 0;
+	//uint8_t choice;
+	atm_state_t next_state = ATM_STATE_IDLE;
+
+	while (1) {
+
+		switch (next_state)
+		{
+			case ATM_STATE_IDLE:
+
+				if (is_card_inserted())
+					next_state = ATM_STATE_CARD_INSERTED;
+
+				break;
+
+			case ATM_STATE_CARD_INSERTED:
+				
+				if (ATM_SUCCESS == pin_check())
+					next_state = ATM_STATE_PIN_CHECK;
```

Comment:
I don't see `ATM_STATE_PIN_CHECK ` state anywhere. This will be stuck here forever if pin checks out